core/src/main/java/org/apache/iceberg/DistributedDataBatchScan.java

```diff
@@ -0,0 +1,309 @@
+    return TableScanUtil.planTaskGroups(
+        planFiles(), targetSplitSize(), splitLookback(), splitOpenFileCost());
+  }
+
+  private CloseableIterable<ScanTask> createTasks(
+      CloseableIterable<DataFile> dataFiles, DeleteFileIndex deletes) {
+
+    String schemaString = SchemaParser.toJson(tableSchema());
+    LoadingCache<Integer, String> specStringCache = specCache(PartitionSpecParser::toJson);
+    LoadingCache<Integer, ResidualEvaluator> residualCache = specCache(this::newResidualEvaluator);
+
+    return CloseableIterable.transform(
+        dataFiles,
+        dataFile -> {
+          DeleteFile[] deleteFiles = deletes.forDataFile(dataFile);
+
+          scanMetrics().totalFileSizeInBytes().increment(dataFile.fileSizeInBytes());
+          scanMetrics().resultDataFiles().increment();
+          scanMetrics().resultDeleteFiles().increment(deleteFiles.length);
+
+          for (DeleteFile deleteFile : deleteFiles) {
+            scanMetrics().totalDeleteFileSizeInBytes().increment(deleteFile.fileSizeInBytes());
+          }
+
+          String specString = specStringCache.get(dataFile.specId());
+          ResidualEvaluator residuals = residualCache.get(dataFile.specId());
+
+          return new BaseFileScanTask(dataFile, deleteFiles, schemaString, specString, residuals);
+        });
+  }
+
+  private List<ManifestFile> findMatchingDataManifests(Snapshot snapshot) {
+    List<ManifestFile> dataManifests = snapshot.dataManifests(io());
+    scanMetrics().totalDataManifests().increment(dataManifests.size());
+
+    List<ManifestFile> matchingDataManifests = filterManifests(dataManifests);
+    int skippedDataManifestsCount = dataManifests.size() - matchingDataManifests.size();
+    scanMetrics().skippedDataManifests().increment(skippedDataManifestsCount);
+
+    return matchingDataManifests;
+  }
+
+  private List<ManifestFile> findMatchingDeleteManifests(Snapshot snapshot) {
+    List<ManifestFile> deleteManifests = snapshot.deleteManifests(io());
+    scanMetrics().totalDeleteManifests().increment(deleteManifests.size());
+
+    List<ManifestFile> matchingDeleteManifests = filterManifests(deleteManifests);
+    int skippedDeleteManifestsCount = deleteManifests.size() - matchingDeleteManifests.size();
+    scanMetrics().skippedDeleteManifests().increment(skippedDeleteManifestsCount);
+
+    return matchingDeleteManifests;
+  }
+
+  private List<ManifestFile> filterManifests(List<ManifestFile> manifests) {
+    LoadingCache<Integer, ManifestEvaluator> evalCache = specCache(this::newManifestEvaluator);
+
+    return manifests.stream()
+        .filter(manifest -> manifest.hasAddedFiles() || manifest.hasExistingFiles())
+        .filter(manifest -> evalCache.get(manifest.partitionSpecId()).eval(manifest))
+        .collect(Collectors.toList());
+  }
+
+  private boolean shouldPlanRemotely(PlanningMode mode, List<ManifestFile> manifests) {
+    switch (mode) {
+      case LOCAL:
```

Comment:
Should this be used if `planExecutor` was set?

Reply:
Maybe? I was planning to use the passed/default executor to assign deletes to data files concurrently. It is probably reasonable to assume someone needs local planning if they pass a custom executor. I'll change it.

Reply:
Changed.